### PR TITLE
Pass original type while creating implicit interface

### LINKED ISSUE #1233: [LAPACK]: Pass old_type while creating implicit interface.
While scanning LAPACK
For
```fortran
      SUBROUTINE SSYR2(UPLO)
      CHARACTER UPLO
      INTEGER INFO
      LOGICAL LSAME
      EXTERNAL LSAME
      INFO = 0
      IF (.NOT.LSAME(UPLO,'U') .AND. .NOT.LSAME(UPLO,'L')) THEN
          INFO = 1
      ELSE 
          INFO = 2
      END IF

      RETURN
      END
```
> LFortran
```console
$ lfortran /home/pranavchiku/lapack/BLAS/SRC/temp.f --show-asr --implicit-typing --implicit-interface --fixed-form --no-warnings --indent
semantic error: Operand of .not. operator is real
 --> /home/pranavchiku/lapack/BLAS/SRC/temp.f:8:11
  |
8 |       IF (.NOT.LSAME(UPLO,'U') .AND. .NOT.LSAME(UPLO,'L')) THEN
  |           ^^^^^^^^^^^^^^^^^^^^ 


Note: if any of the above error or warning messages are not clear or are lacking
context please report it to us (we consider that a bug that needs to be fixed).
```

This is because we are currently hardcoding the return type to real-8 and hence it [LSAME](https://netlib.org/lapack/explore-3.1.1-html/lsame.f.html) returns `real`.

### PR DESCRIPTION
**Pull request by Pranavchiku**
Fixes #1233.

### CODE REVIEW

File: src/lfortran/semantics/ast_common_visitor.h [hunk 1]
```diff
@@ -3187,8 +3185,13 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
...
         // FIXME: accept this type as an argument
         // currently hardcoding the return type to real-8
-        ASR::ttype_t *type = ASRUtils::TYPE(ASR::make_Real_t(al, x.base.base.loc,
+        ASR::ttype_t *type;
+        if ( old_type == nullptr ) {
+            type = ASRUtils::TYPE(ASR::make_Real_t(al, x.base.base.loc,
                                 8, nullptr, 0));
+        } else {
+            type = old_type;
+        }
```
Comment (certik) [inline src/lfortran/semantics/ast_common_visitor.h:3194]: Why not simply always accept the type as an argument (non-null)?

Comment (Pranavchiku) [reply src/lfortran/semantics/ast_common_visitor.h:3194]: We can, let me apply that.

File: src/lfortran/semantics/ast_common_visitor.h [hunk 2]
```diff
@@ -3323,7 +3318,10 @@ class CommonVisitor : public AST::BaseVisitor<Derived> {
                 // Function Call is not defined in this case.
                 // We need to create an interface and add the Function into
                 // the symbol table.
-                create_implicit_interface_function(x, var_name, true);
+                // Currently using real*8 as the return type.
+                ASR::ttype_t* type = ASRUtils::TYPE(ASR::make_Real_t(al, x.base.base.loc,
```
Comment (Pranavchiku) [inline src/lfortran/semantics/ast_common_visitor.h:3322]: Here, function is not present in symbol table and hence we have no idea about return type of it, so using `real*8` in this case.

File: src/lfortran/semantics/ast_body_visitor.cpp [hunk 3]
```diff
@@ -1750,7 +1750,9 @@ class BodyVisitor : public CommonVisitor<BodyVisitor> {
         if (!original_sym) {
             original_sym = resolve_intrinsic_function(x.base.base.loc, sub_name);
             if (!original_sym && compiler_options.implicit_interface) {
-                create_implicit_interface_function(x, sub_name, false);
+                ASR::ttype_t* type = ASRUtils::TYPE(ASR::make_Real_t(al, x.base.base.loc,
```
Comment (Pranavchiku) [inline src/lfortran/semantics/ast_body_visitor.cpp:1753]: Same explanation as in previous comment.

**Review by certik — APPROVED** [review]
I think this looks good. Make sure to rebase or squash before merging.

**Pull request merged**